### PR TITLE
Disable _all by default, disallow configuring _all on 6.0+ indices

### LINKED ISSUE #19784: Explore disabling the _all field by default
We should explore the idea of disabling the `_all` field by default. For
background, the `_all` field contains the contents of each of a document's
field. During indexing, these contents are copied to the `_all` field, analyzed
with a specific analyzer, and then indexed. At query time, some queries such as
the `query_string` and `simple_query_string` queries search the `_all` field if
no fields are specified.

There are a number of issues with the `_all` field:
- It uses a fair amount of disk space, due to duplicating the values from all
  other fields to an additional field
- `_all` has its own analyzer, which is confusing when expecting to query
  text analyzed a certain way (or with synonyms, for example) and discovering
  that it does not match due to a analysis difference
- Additional indexing overhead caused by data duplication
- Since the `_all` field is not retrievable or part of `_source`, its contents
  cannot easily be inspected for debugging purposes, some users do not know it
  even exists, which causes confusion at query time
- Better alternatives exist, such as the `copy_to` value on mappings which can
  be used to create custom `_all` fields

If we were to change the `_all` field to be disabled by default, some queries
would have to be handled differently. The `query_string` and
`simple_query_string` queries would have to know a default field or fields to
query. Perhaps we would be able to change these queries to send a `"fields":
["*"]` parameter if no fields are specified in the JSON query, so that all field
values can still be queried.

### PR DESCRIPTION
**Pull request by dakrone**
This change disables the `_all` meta field by default.

Now that we have the "all-fields" method of query execution, we can save both
indexing time and disk space by disabling it.

`_all` can no longer be configured for indices created after 6.0.

Relates to #20925 and #21341
Resolves #19784

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java [hunk 1]
```diff
@@ -2435,6 +2435,7 @@ public void testPostingsHighlighterCommonTermsQuery() throws IOException {
 
     private static XContentBuilder type1PostingsffsetsMapping() throws IOException {
         return XContentFactory.jsonBuilder().startObject().startObject("type1")
+                .startObject("_all").field("enabled", true).endObject()
                 .startObject("properties")
                 .startObject("field1").field("type", "text").field("index_options", "offsets").endObject()
                 .startObject("field2").field("type", "text").field("index_options", "offsets").endObject()
```
Comment (jimczi) [inline core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java]: Can you explain very briefly why it fails the test with _all disabled ?

File: core/src/test/java/org/elasticsearch/search/nested/SimpleNestedIT.java [hunk 2]
```diff
@@ -56,7 +56,9 @@
 
 public class SimpleNestedIT extends ESIntegTestCase {
     public void testSimpleNested() throws Exception {
-        assertAcked(prepareCreate("test").addMapping("type1", "nested1", "type=nested").addMapping("type2", "nested1", "type=nested"));
+        assertAcked(prepareCreate("test")
+                .addMapping("type1", "nested1", "type=nested", "_all", "enabled=true")
+                .addMapping("type2", "nested1", "type=nested", "_all", "enabled=true"));
         ensureGreen();
 
         // check on no data, see it works
```
Comment (jimczi) [inline core/src/test/java/org/elasticsearch/search/nested/SimpleNestedIT.java]: And here ;)

File: core/src/test/java/org/elasticsearch/search/query/SimpleQueryStringIT.java [hunk 3]
```diff
@@ -252,8 +252,7 @@ public void testSimpleQueryStringFlags() throws ExecutionException, InterruptedE
         searchResponse = client()
                 .prepareSearch()
                 .setQuery(
-                        simpleQueryStringQuery("baz | egg*").defaultOperator(Operator.AND).flags(SimpleQueryStringFlag.WHITESPACE,
-                                SimpleQueryStringFlag.PREFIX)).get();
+                        simpleQueryStringQuery("blah | egg*").flags(SimpleQueryStringFlag.WHITESPACE, SimpleQueryStringFlag.PREFIX)).get();
```
Comment (jimczi) [inline core/src/test/java/org/elasticsearch/search/query/SimpleQueryStringIT.java]: Why do you need to remove this `.defaultOperator(Operator.AND)`. It looks like a test use case.

File: docs/reference/migration/migrate_6_0/mappings.asciidoc [hunk 4]
```diff
@@ -1,6 +1,15 @@
 [[breaking_60_mappings_changes]]
 === Mapping changes
 
+==== The `_all` meta field is now disabled by default
+
+On new mappings, the `_all` meta field that contains a copy of the text from
+each field is not disabled by default. The `query_string` and
```
Comment (jimczi) [inline docs/reference/migration/migrate_6_0/mappings.asciidoc]: is not enabled by default ?

File: docs/reference/migration/migrate_6_0/mappings.asciidoc [hunk 5]
```diff
@@ -1,6 +1,15 @@
...
+`simple_query_string` queries that previously used `_all` to search will now
+check if `_all` is enabled/disabled and switch to executing the query across all
+fields if `_all` is disabled. `_all` may be re-enabled by setting `enabled`:
+`true` in the mapping configuration for `_all`.
+
```
Comment (jimczi) [inline docs/reference/migration/migrate_6_0/mappings.asciidoc]: Maybe add a sentence that says that a reindex is required when `_all` is re-enabled ?

File: docs/reference/search/validate.asciidoc [hunk 6]
```diff
@@ -87,7 +87,15 @@ due to dynamic mapping, and 'foo' does not correctly parse into a date:
 
 [source,js]
 --------------------------------------------------
-GET twitter/tweet/_validate/query?q=post_date:foo%5d
+GET twitter/tweet/_validate/query
+{
+  "query": {
+    "query_string": {
+      "query": "post_date:foo",
+      "lenient": false
```
Comment (jimczi) [inline docs/reference/search/validate.asciidoc]: `false` is still the default value, so why is this needed ?

File: docs/reference/search/validate.asciidoc [hunk 7]
```diff
@@ -102,7 +110,15 @@ about why a query failed:
...
 [source,js]
 --------------------------------------------------
-GET twitter/tweet/_validate/query?q=post_date:foo%5d&explain=true
+GET twitter/tweet/_validate/query?explain=true
+{
+  "query": {
+    "query_string": {
+      "query": "post_date:foo",
+      "lenient": false
+    }
```
Comment (jimczi) [inline docs/reference/search/validate.asciidoc]: same here

**Review by jimczi — COMMENTED** [review]
Just left some minor things  that need clarifications. 
Other than that this change looks good. I am surprised that it did not require more changes in the tests but I guess that the `all_fields` mode covers most of the use case.

File: core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java (hunk 1 above)
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java]: `testPostingsHighlighter` has a test that queries on the `_all` field and highlights other fields. Without this, it would fail.

I can remove the particular clause if you think that would be a better solution?

File: core/src/test/java/org/elasticsearch/search/nested/SimpleNestedIT.java (hunk 2 above)
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/search/nested/SimpleNestedIT.java]: There is an actual check, see a few lines down for:

```
// check that _all is working on nested docs
```

Which requires `_all` for the test to work.

File: core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java (hunk 1 above)
Comment (jimczi) [inline core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java]: Ok thanks for the explanation. BTW are we planning to add support for `all_fields` in the multi match query ?

File: core/src/test/java/org/elasticsearch/search/nested/SimpleNestedIT.java (hunk 2 above)
Comment (jimczi) [inline core/src/test/java/org/elasticsearch/search/nested/SimpleNestedIT.java]: Makes sense. Thanks

File: core/src/test/java/org/elasticsearch/search/query/SimpleQueryStringIT.java (hunk 3 above)
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/search/query/SimpleQueryStringIT.java]: I'll change this to be a bit more transparent of a test. It's not a very good check right now.

File: docs/reference/search/validate.asciidoc (hunk 7 above)
Comment (dakrone) [inline docs/reference/search/validate.asciidoc]: Same as above :)

**Comment by dakrone** [comment]
I pushed some commits, thanks for looking @jimczi!

**Comment by jpountz** [comment]
I don't like having a feature that has such a large impact on the code base disabled by default, as it means the maintenance cost remains the same (=high), while the benefit to our users decreases significantly, as most of them will not opt in. Two possible paths that I can think of are either to remove it entirely (ie. forbid 6.x indices from enabling it and removing in 7), or keep `_all` as a simple `text` field that is used as a name convention for a catch-all field that other fields can copy to.

**Comment by jimczi** [comment]
> I don't like having a feature that has such a large impact on the code base disabled by default

The impact is pretty contained and should not change the OOB significantly. Though I agree that keeping the feature is costly so I think that forbids in 6 and removing in 7 is a good plan. 

> or keep _all as a simple text field that is used as a name convention for a catch-all field that other fields can copy to.

Not sure it would be useful if we don't copy to it by default and if we always use the `all_fields` mode in the `query_string`. It's really an OOB issue which seems to be resolved by the `all_fields` mode.

**Comment by dakrone** [comment]
Alright, sounds like the consensus is to disallow in 6.0+ and we can eventually remove it in 7.0, I'll update the PR and title.

**Comment by dakrone** [comment]
@jpountz @jimczi okay, I've changed this to disallow configuring `_all` for indices created 6.0+ in addition to it being disabled by default. Sorry for the large PR, but it was mostly removing tests that are no longer needed.

**Comment by dakrone** [comment]
retest this please

File: docs/reference/migration/migrate_6_0/mappings.asciidoc [hunk 8]
```diff
@@ -1,6 +1,17 @@
 [[breaking_60_mappings_changes]]
 === Mapping changes
 
+==== The `_all` meta field is now disabled by default
+
+On new mappings, the `_all` meta field that contains a copy of the text from
+each field is now disabled by default. The `query_string` and
+`simple_query_string` queries that previously used `_all` to search will now
+check if `_all` is enabled/disabled and switch to executing the query across all
+fields if `_all` is disabled. `_all` may be re-enabled by setting `enabled`:
```
Comment (vincentfretin) [inline docs/reference/migration/migrate_6_0/mappings.asciidoc]: > switch to executing the query across all fields if `_all` is disabled

I hope not all fields, but only the fields that I specified the include_in_all: True option? Can you please clarify?

File: core/src/test/java/org/elasticsearch/action/admin/indices/template/BWCTemplateTests.java [hunk 9]
```diff
@@ -1,55 +0,0 @@
...
-    }
-
-    public void testLogstashTemplatesBWC() throws Exception {
-        String ls5x = copyToStringFromClasspath("/org/elasticsearch/action/admin/indices/template/logstash-5.0.template.json");
-        client().admin().indices().preparePutTemplate("logstash-5x").setSource(ls5x).get();
-        client().prepareIndex("logstash-foo", "doc", "1").setSource("message", "foo").get();
-        assertWarnings("Deprecated field [template] used, replaced by [index_patterns]");
-    }
-
-}
```
Comment (jimczi) [inline core/src/test/java/org/elasticsearch/action/admin/indices/template/BWCTemplateTests.java]: We can maybe keep this test around and just remove _all from the templates ?

File: docs/reference/migration/migrate_6_0/mappings.asciidoc (hunk 8 above)
Comment (dakrone) [inline docs/reference/migration/migrate_6_0/mappings.asciidoc]: If you don't specify anything in the `fields` parameter, then `simple_query_string` and `query_string` will automatically try to execute the search across all searchable fields (see: https://github.com/elastic/elasticsearch/pull/20925). If you want to limit it, then you can specify the `"fields"` parameter in the query.

Additionally, you could use `copy_to` and create your own all-type field as a composite of whatever fields you want, and then set that as the `default_field` for the index.

Comment (vincentfretin) [inline docs/reference/migration/migrate_6_0/mappings.asciidoc]: Thanks for your prompt response and the link to #20925. Currently in elasticsearch 5.0 I use "include_in_all: False" on everything except on text fields, so what I'll have to do with this change is indeed to create my own field with copy_to, or specify all the text fields I want in fields when I do the query.

File: docs/reference/migration/migrate_6_0/mappings.asciidoc [hunk 10]
```diff
@@ -1,6 +1,17 @@
...
+`true` in the mapping configuration for `_all`. If `_all` is enabled on an
+existing index data that was previously indexed will need to be reindexed in
+order to populate the `_all` field.
+
```
Comment (jimczi) [inline docs/reference/migration/migrate_6_0/mappings.asciidoc]: This is complicated and maybe not needed. If we don't accept new mapping with _all enabled maybe we should also not allow to enable `_all` on an index that was created in older versions ? So we would completely disallow enabling `_all` but continue to handle indices created in older versions with `_all` already indexed ?

**Review by jimczi — COMMENTED** [review]
Thanks @dakrone 
This looks good. I can imagine how painful it was to understand why all this tests were relying on _all and how to remove it without loosing the test case.
Though I think we should at least keep some test around regarding the AllFieldType and maybe the copyToAll functionality. It will surely be removed but in the mean time we may add things that breaks it. 

IMO we should completely disallow enabling `_all` even if the index was created < 6.x. We don't want to create more `_all` users ;)

File: core/src/test/java/org/elasticsearch/action/admin/indices/template/BWCTemplateTests.java (hunk 9 above)
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/action/admin/indices/template/BWCTemplateTests.java]: Sure, though then it won't be the real templates from Logstash and Beats, but I guess it is still testing the other features.

File: docs/reference/migration/migrate_6_0/mappings.asciidoc (hunk 10 above)
Comment (dakrone) [inline docs/reference/migration/migrate_6_0/mappings.asciidoc]: We do disallow this for 5.0 indices, see `UpdateMappingOnClusterIT` which tests this

Comment (jimczi) [inline docs/reference/migration/migrate_6_0/mappings.asciidoc]: Ok good then you can remove this sentence ?

Comment (dakrone) [inline docs/reference/migration/migrate_6_0/mappings.asciidoc]: Definitely, good catch!

**Comment by dakrone** [comment]
@jimczi I re-added some tests based on your feedback, thanks for taking a look!

> regarding the AllFieldType and maybe the copyToAll functionality

Can you describe what you mean here? Do you mean just that fields are copied to the `_all` field automatically?

**Comment by jimczi** [comment]
> Can you describe what you mean here? Do you mean just that fields are copied to the _all field automatically?

Yes and the AllFieldTypeTest. Why is it removed ?

**Comment by dakrone** [comment]
> the AllFieldTypeTest. Why is it removed ?

My accidental over-zealousness, I've re-added it :)

As far as testing that fields are copied, the `_all` field is still used for the query_string and simple_query_string tests, which tests that the fields are copied to the `_all` field (since it queries on them)

**Comment by dakrone** [comment]
retest this please

**Review by jimczi — APPROVED** [review]
LGTM

**Comment by lukas-vlcek** [comment]
If I want to index documents that have `_all` field in the top level would it be possible to configure `_all` field in mapping? AFAIK it is currently not possible to index documents having `_all` field but in the future, when the historical function of `_all` field is removed, wouldn't it make sense to threat this field as any other document field?

In other words, will it be possible to index documents like the following one OOB?

````json
{
  "foo": "bar",
  "_all": "dummy"
}
````

**Comment by dakrone** [comment]
> In other words, will it be possible to index documents like the following one OOB?

We strongly recommend not naming user-created fields started with an `_`, as those are reserved for internal use.

You could, however, name it `my_all` (or something else) and use the `copy_to` feature on all other mapped fields to create your own version of an `_all` field.
